From a592e39d7bd40a2c9af25091977483793597eb06 Mon Sep 17 00:00:00 2001
From: De Huo <De.Huo@windriver.com>
Date: Fri, 19 Jul 2019 15:40:19 +0800
Subject: [PATCH] Add rule accounts_passwords_pam_tally2_deny instead 
to WRLinux1019 profile

Add rule accounts_passwords_pam_tally2_deny instead of
 accounts_passwords_pam_faillock_deny to WRLinux1019 profile to algin with
 WRLinux1019 product

Upstream-Status: Pending

Signed-off-by: De Huo <De.Huo@windriver.com>
---
 .../ansible/shared.yml                             | 81 ++++++++++++++++++++++
 .../bash/wrlinux1019.sh                            | 16 +++++
 .../oval/wrlinux1019.xml                           | 63 +++++++++++++++++
 .../accounts_passwords_pam_tally2_deny/rule.yml    | 60 ++++++++++++++++
 .../profiles/draft_stig_wrlinux_disa.profile       |  1 +
 5 files changed, 221 insertions(+)
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny/ansible/shared.yml
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny/bash/wrlinux1019.sh
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny/oval/wrlinux1019.xml
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny/rule.yml

diff --git a/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny/ansible/shared.yml b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny/ansible/shared.yml
new file mode 100644
index 0000000..d2b08c0
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny/ansible/shared.yml
@@ -0,0 +1,81 @@
+# platform = multi_platform_rhel,multi_platform_fedora,multi_platform_ol,multi_platform_rhv
+# reboot = false
+# strategy = restrict
+# complexity = low
+# disruption = low
+- (xccdf-var var_accounts_passwords_pam_faillock_deny)
+
+- name: Add auth pam_faillock preauth deny before pam_unix.so
+  pamd:
+    name: "{{ item }}"
+    type: auth
+    control: sufficient
+    module_path: pam_unix.so
+    new_type: auth
+    new_control: required
+    new_module_path: pam_faillock.so
+    module_arguments: 'preauth
+        silent
+        deny={{ var_accounts_passwords_pam_faillock_deny }}'
+    state: before
+  loop:
+    - system-auth
+    - password-auth
+
+- name: Add deny argument to auth pam_faillock preauth
+  pamd:
+    name: "{{ item }}"
+    type: auth
+    control: required
+    module_path: pam_faillock.so
+    module_arguments: 'preauth
+        silent
+        deny={{ var_accounts_passwords_pam_faillock_deny }}'
+    state: args_present
+  loop:
+    - system-auth
+    - password-auth
+
+- name: Add auth pam_faillock authfail deny after pam_unix.so
+  pamd:
+    name: "{{ item }}"
+    type: auth
+    control: sufficient
+    module_path: pam_unix.so
+    new_type: auth
+    new_control: '[default=die]'
+    new_module_path: pam_faillock.so
+    module_arguments: 'authfail
+        deny={{ var_accounts_passwords_pam_faillock_deny }}'
+    state: after
+  loop:
+    - system-auth
+    - password-auth
+
+- name: Add deny argument to auth pam_faillock authfail
+  pamd:
+    name: "{{ item }}"
+    type: auth
+    new_type: auth
+    control: '[default=die]'
+    module_path: pam_faillock.so
+    module_arguments: 'authfail
+        deny={{ var_accounts_passwords_pam_faillock_deny }}'
+    state: args_present
+  loop:
+    - system-auth
+    - password-auth
+
+- name: Add account pam_faillock before pam_unix.so
+  pamd:
+    name: "{{ item }}"
+    type: account
+    control: required
+    module_path: pam_unix.so
+    new_type: account
+    new_control: required
+    new_module_path: pam_faillock.so
+    state: before
+  loop:
+    - system-auth
+    - password-auth
diff --git a/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny/bash/wrlinux1019.sh b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny/bash/wrlinux1019.sh
new file mode 100644
index 0000000..6a51638
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny/bash/wrlinux1019.sh
@@ -0,0 +1,16 @@
+# platform = multi_platform_wrlinux
+
+. /usr/share/scap-security-guide/remediation_functions
+populate var_accounts_passwords_pam_faillock_deny
+
+if grep -q '^[^#].*pam_tally2.so' /etc/pam.d/common-auth; then
+  if grep -q 'deny=' /etc/pam.d/common-auth; then
+	  sed -i "s/deny=[0-9]*/deny=$var_accounts_passwords_pam_faillock_deny/" /etc/pam.d/common-auth
+  else
+	  sed -i "/auth.*pam_tally2.so/s/$/ deny=$var_accounts_passwords_pam_faillock_deny/" /etc/pam.d/common-auth
+  fi
+else
+  echo "auth  required pam_tally2.so deny=$var_accounts_passwords_pam_faillock_deny" >>/etc/pam.d/common-auth
+fi
+
+echo "account	required	pam_tally2.so" >>/etc/pam.d/common-account
diff --git a/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny/oval/wrlinux1019.xml b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny/oval/wrlinux1019.xml
new file mode 100644
index 0000000..9a185e4
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny/oval/wrlinux1019.xml
@@ -0,0 +1,63 @@
+<def-group>
+  <definition class="compliance" id="accounts_passwords_pam_tally2_deny" version="1">
+    <metadata>
+      <title>Lock out account after failed login attempts</title>
+      <affected family="unix">
+        <platform>multi_platform_all</platform>
+      </affected>
+      <description>The number of allowed failed logins should be set correctly.</description>
+    </metadata>
+    <criteria operator="AND" comment="Checks common to both scenarios">
+
+      <criterion test_ref="test_accounts_passwords_pam_tally2_common-auth"
+      comment="pam_tally2.so deny set in common-auth" />
+      <criterion test_ref="test_accounts_passwords_pam_tally2_common-account"
+      comment="pam_tally2.so set in common-account" />
+
+    </criteria>
+  </definition>
+
+  <!-- Specify required external variable & create corresponding state from it -->
+  <external_variable id="var_accounts_passwords_pam_faillock_deny" datatype="int"
+  comment="number of failed login attempts allowed" version="1" />
+
+  <ind:textfilecontent54_state id="state_var_accounts_passwords_pam_tally2_deny_value" version="1">
+    <ind:subexpression datatype="int" operation="less than or equal" var_ref="var_accounts_passwords_pam_faillock_deny" />
+  </ind:textfilecontent54_state>
+
+  <!-- Check for preauth silent in /etc/pam.d/common-auth -->
+  <!-- Also check the 'deny' option value matches the number of failed login attempts allowed -->
+  <ind:textfilecontent54_test id="test_accounts_passwords_pam_tally2_common-auth"
+  check="all" check_existence="all_exist"
+  comment="Check pam_tally2.so deny in /etc/pam.d/common-auth" version="1">
+    <ind:object object_ref="object_accounts_passwords_pam_tally2_deny_common-auth" />
+    <ind:state state_ref="state_var_accounts_passwords_pam_tally2_deny_value" />
+  </ind:textfilecontent54_test>
+
+  <ind:textfilecontent54_object id="object_accounts_passwords_pam_tally2_deny_common-auth" version="1">
+    <!-- Read whole /etc/pam.d/common-auth content as single line so we can verify existing order of PAM modules -->
+    <ind:behaviors singleline="true" />
+    <ind:filepath>/etc/pam.d/common-auth</ind:filepath>
+    <!-- Since order of PAM modules matters ensure pam_tally2.so deny in auth section is listed before
+         pam_unix.so module in auth section -->
+    <ind:pattern operation="pattern match">[\n][\s]*auth[\s]+(?:(?:sufficient)|(?:\[.*default=ignore.*\]))[\s]+pam_unix\.so[^\n]*[\n][\s\S]*[\n][\s]*auth[\s]+required[\s]+pam_tally2\.so[\s]+deny=([0-9]+).*</ind:pattern>
+    <!-- Check only the first instance -->
+    <ind:instance datatype="int" operation="equals">1</ind:instance>
+  </ind:textfilecontent54_object>
+
+  <!-- Check for pam_tally2.so present in /etc/pam.d/common-account -->
+  <ind:textfilecontent54_test id="test_accounts_passwords_pam_tally2_common-account"
+  check="all" check_existence="all_exist"
+  comment="Check if pam_tally2.so is called in /etc/pam.d/common-account" version="1" >
+    <ind:object object_ref="object_accounts_passwords_pam_tally2_common-account" />
+  </ind:textfilecontent54_test>
+
+  <ind:textfilecontent54_object id="object_accounts_passwords_pam_tally2_common-account" version="1">
+    <!-- Read whole /etc/pam.d/common-account content as single line so we can verify existing order of PAM modules -->
+    <ind:filepath>/etc/pam.d/common-account</ind:filepath>
+    <ind:pattern operation="pattern match">[^#]*account[\s]+required[\s]+pam_tally2\.so.*</ind:pattern>
+    <!-- Check only the first instance -->
+    <ind:instance datatype="int" operation="equals">1</ind:instance>
+  </ind:textfilecontent54_object>
+
+</def-group>
diff --git a/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny/rule.yml b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny/rule.yml
new file mode 100644
index 0000000..29c18eb
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny/rule.yml
@@ -0,0 +1,60 @@
+documentation_complete: true
+
+prodtype: wrlinux1019,rhel6,rhel7,rhel8,fedora,ol7,ol8,rhv4
+
+title: 'Set Deny For Failed Password Attempts'
+
+description: |-
+    To configure the system to lock out accounts after a number of incorrect login
+    attempts using <tt>pam_tally2.so</tt>, modify the content of both
+    <tt>/etc/pam.d/common-auth</tt> and <tt>/etc/pam.d/common-account</tt> as follows:
+    <br /><br />
+    <ul>
+    <li> add the following line immediately <tt>before</tt> the <tt>pam_unix.so</tt> statement in the <tt>AUTH</tt> section:
+    <pre>auth required pam_tally2.so preauth silent deny=<sub idref="var_accounts_passwords_pam_tally2_deny" /> unlock_time=<sub idref="var_accounts_passwords_pam_tally2_deny" /> fail_interval=<sub idref="var_accounts_passwords_pam_faillock_fail_interval" /></pre></li>
+    <li> add the following line immediately <tt>after</tt> the <tt>pam_unix.so</tt> statement in the <tt>AUTH</tt> section:
+    <pre>auth [default=die] pam_tally2.so authfail deny=<sub idref="var_accounts_passwords_pam_tally2_deny" /> unlock_time=<sub idref="var_accounts_passwords_pam_faillock_unlock_time" /> fail_interval=<sub idref="var_accounts_passwords_pam_faillock_fail_interval" /></pre></li>
+    <li> add the following line immediately <tt>before</tt> the <tt>pam_unix.so</tt> statement in the <tt>ACCOUNT</tt> section:
+    <pre>account required pam_tally2.so</pre></li>
+    </ul>
+
+rationale: |-
+    Locking out user accounts after a number of incorrect attempts
+    prevents direct password guessing attacks.
+
+severity: medium
+
+identifiers:
+    cce@rhel6: 26844-1
+    cce@rhel7: 27350-8
+    cce@rhel8: 80667-9
+
+references:
+    stigid@rhel6: RHEL-06-000061
+    srg@rhel6: SRG-OS-000021
+    disa@rhel6: '44'
+    cis: 5.3.2
+    cjis: 5.5.3
+    cui: 3.1.8
+    disa: "2238,44"
+    nist: AC-7(a)
+    nist-csf: PR.AC-7
+    ospp: FMT_MOF_EXT.1
+    pcidss: Req-8.1.6
+    srg: SRG-OS-000329-GPOS-00128,SRG-OS-000021-GPOS-00005
+    vmmsrg: SRG-OS-000021-VMM-000050
+    stigid@rhel7: "010320"
+    isa-62443-2013: 'SR 1.1,SR 1.10,SR 1.2,SR 1.5,SR 1.7,SR 1.8,SR 1.9'
+    isa-62443-2009: 4.3.3.6.1,4.3.3.6.2,4.3.3.6.3,4.3.3.6.4,4.3.3.6.5,4.3.3.6.6,4.3.3.6.7,4.3.3.6.8,4.3.3.6.9
+    cobit5: DSS05.04,DSS05.10,DSS06.10
+    iso27001-2013: A.18.1.4,A.9.2.1,A.9.2.4,A.9.3.1,A.9.4.2,A.9.4.3
+    cis-csc: 1,12,15,16
+
+ocil_clause: 'that is not the case'
+
+ocil: |-
+    To ensure the failed password attempt policy is configured correctly, run the following command:
+    <pre>$ grep pam_tally2 /etc/pam.d/system-auth</pre>
+    The output should show <tt>deny=<sub idref="var_accounts_passwords_pam_tally2_deny" /></tt>.
+
+platform: pam
diff --git a/wrlinux1019/profiles/draft_stig_wrlinux_disa.profile b/wrlinux1019/profiles/draft_stig_wrlinux_disa.profile
index b2c2a48..0efec58 100644
--- a/wrlinux1019/profiles/draft_stig_wrlinux_disa.profile
+++ b/wrlinux1019/profiles/draft_stig_wrlinux_disa.profile
@@ -230,4 +230,5 @@ selections:
     - audit_rules_usergroup_modification
     - accounts_passwords_pam_faillock_deny
     - accounts_passwords_pam_faillock_deny_root
+    - accounts_passwords_pam_tally2_deny
     - package_openssh-server_installed
-- 
1.9.1

